core/src/main/java/org/elasticsearch/common/io/Streams.java

```diff
@@ -236,4 +241,57 @@ public static void readAllLines(InputStream input, Callback<String> callback) th
             }
         }
     }
+
+    /**
+     * Wraps the given {@link BytesStreamOutput} in a {@link StreamOutput} that simply flushes when
+     * close is called. The returned values from this method also implements the {@link BytesStream}
+     * interface.
+     */
+    public static StreamOutput flushOnCloseStream(BytesStreamOutput os) {
+        return new FlushOnCloseOutputStream(os);
+    }
+
+    /**
+     * A wrapper around a {@link BytesStreamOutput} that makes the close operation a flush. This is
+     * needed as sometimes a stream will be closed but the bytes that the stream holds still need
+     * to be used and the stream cannot be closed until the bytes have been consumed.
+     */
+    private static class FlushOnCloseOutputStream extends StreamOutput implements BytesStream {
```

Comment:
why does this implement BytesStream? from the interface of the factory method it's unneeded? I also wonder if we should simply subclass `OutputStreamStreamOutput`  here and just require `StreamOutput` as the input interface